README.md

```diff
@@ -1,73 +1,320 @@
+| public_key_content   | Contains the public key used by this host for public key authentication without certificates.                        | str         | no       | -                       | leftrsasigkey/rightrsasigkey |
+| cert_name            | Defines the certificate nickname of this host's certificate in the NSS database.                  | str         | no       | -                       | leftcert/rightcert   |
+| external             | Indicates that the associated host is external to the inventory (e.g. not managed by our inventory).    | bool        | no       | False                   | -                    |
+| subnets              | Defines a list of the subnets that should be available via the VPN connection.                    | list        | no       | -                       | leftsubnets/rightsubnets |
+
+#### public_key
+
+`public_key_src` specifies a path to a file on the controller host containing the public key used by this host for public key authentication without certificates. Otherwise, the user can directly specify the public key for this host by populating `public_key_content`. `public_key_content` can also accept a CKAID or nickname for a public key in the NSS database.
+
+Note that `public_key_src` and `public_key_content` may also be specified as host variables in the Ansible `all.hosts` section. The variable names in this case will be `vpn_public_key_src` and `vpn_public_key_content`. See [this example](#defining-public-keys-at-different-levels).
+
+**To be added later**: if neither `public_key_src` nor `public_key_content` is populated, the role will generate key pairs for each host.
+
+### Handling special cases with host names
+
+There are two ways to handle the special case when the names of the hosts in the Ansible inventory are not FQDN hostnames that you can use directly (e.g. they may be host aliases like "bastion_east"):  
+
+1. **Use the Ansible `hosts` configuration**
 
 ```yaml
-template_foo: "oof"
-template_bar: "baz"
+all:
+  hosts:
+    bastion_east:
+      ansible_host: bastion1.example.com
+      vpn_hostid: '@bastion-east'
+    bastion_west:
+      ansible_host: bastion2.example.com
+      vpn_hostid: '@bastion-west'
+    bastion_north:
+      ansible_host: bastion3.example.com
+      vpn_hostid: '@bastion-north'
+  vars:
+    vpn_connections:
+      - hosts:
+          bastion_east:
+          bastion_west:
+          bastion_north:
 ```
+The role would lookup the value of the host to use from `hostvars[hostname].ansible_host`.  You can specify other variables which are always associated with a particular host, like `vpn_hostid`, in the host configuration, which would be used for every connection where that host is referenced.
 
-### Variables Exported by the Role
+2. **Directly specify the host to use**
```

Comment:
I would say that `host` is a misnomer here. What seems to be intended is what address to use (for multi-homed hosts). Maybe then the `host/vpn_host` attribute should be named `address/vpn_address`, because having a `host` attribute in the list of `hosts` is strange (everything in the list of hosts is a host, after all).

Reply:
@richm can you weigh in on this? I think the reason that we were using `host` language is because there might be an IP address, FQDN, or other friendly name in that attribute, though maybe `address` is still an accurate way to describe all of those things?

Reply:
> What seems to be intended is what address to use (for multi-homed hosts).

That's one of the use cases, but there are other use cases where the machine may have two different host names - one used by ansible, and one used by the vpn.  So we need a word that means a hostname or an IP address.  Any suggestions?

Reply:
or do we need two different fields - one for a hostname, and one for an IP address?

Reply:
I can't think of any better suggestion for a single field name besides `address` (I guess a hostname is sort of like a human-readable address?). I also thought of `destination`, `id`, or `machine`, but don't particularly care for any of those. Otherwise, I think it would be okay to create two different fields (even though it's not explicitly necessary).

Reply:
Is this the hostname or address of the remote node?  If so, what about `remote_node`?

Reply:
Or - perhaps the problem is the word `hosts` in this context? Each item in the `vpn_connections` list could be a
* `connection` or `tunnel` - but this implies a 1:1, where this is really a many:many relationship
* `group` - but this word is quite overloaded
* `tunnel_group`
* `mesh`

Reply:
> Is this the hostname or address of the remote node? If so, what about `remote_node`?

What we need is some term that means "this is the name of the host or ip address to use for the actual vpn connection that may be different than the name of the host used in the Ansible inventory"